pkg/pulumiyaml/packages/file.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024, Pulumi Corporation.  All rights reserved.
+
+// Package packages contains utilities for working with Pulumi package lock files.
+package packages
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/blang/semver"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"gopkg.in/yaml.v3"
+)
+
+type ParameterizationDecl struct {
+	// Name is the name of the parameterized package.
+	Name string `yaml:"name"`
+	// Version is the version of the parameterized package.
+	Version string `yaml:"version"`
+	// Value is the value of the parameter. Value is a base64 encoded byte array, use SetValue and GetValue to manipulate the actual value.
+	Value string `yaml:"value"`
+}
+
+func (p *ParameterizationDecl) GetValue() ([]byte, error) {
+	return base64.StdEncoding.DecodeString(p.Value)
+}
+
+func (p *ParameterizationDecl) SetValue(value []byte) {
+	p.Value = base64.StdEncoding.EncodeToString(value)
+}
+
+type PackageDecl struct {
+	// PackageDeclarationVersion is the version of the package declaration file.
+	PackageDeclarationVersion int `yaml:"packageDeclarationVersion"`
+	// Name is the name of the plugin.
+	Name string `yaml:"name"`
+	// Version is the version of the plugin.
+	Version string `yaml:"version,omitempty"`
+	// PluginDownloadURL is the URL to download the plugin from.
+	DownloadURL string `yaml:"downloadUrl,omitempty"`
+	// Parameterization is the parameterization of the package.
+	Parameterization *ParameterizationDecl `yaml:"parameterization,omitempty"`
+}
+
+func (p *PackageDecl) Valid() bool {
+	// All packages should define their version as 1
+	if p.PackageDeclarationVersion != 1 {
+		return false
+	}
+
+	// All packages need a name
+	if p.Name == "" {
+		return false
+	}
+
+	// If parameterization is not nil, it must be valid.
+	if p.Parameterization != nil {
+		return p.Parameterization.Name != "" && p.Parameterization.Version != ""
+	}
+
+	return true
+}
+
+// SearchPackageDecls searches the given directory down recursively for package lock .yaml files.
+func SearchPackageDecls(directory string) ([]PackageDecl, error) {
+	if directory == "" {
+		directory = "."
+	}
+
+	var packages []PackageDecl
+	err := filepath.WalkDir(directory, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// If the file is a directory, skip it.
+		if d.IsDir() {
+			return nil
+		}
+
+		// If the file is not a .yaml file, skip it.
+		if filepath.Ext(path) != ".yaml" {
+			return nil
+		}
+
+		// Read the file.
+		data, err := os.ReadFile(path)
+		if err != nil {
+			return fmt.Errorf("reading %s: %w", path, err)
+		}
+
+		var packageDecl PackageDecl
+		if err := yaml.Unmarshal(data, &packageDecl); err != nil {
+			return fmt.Errorf("unmarshalling %s: %w", path, err)
+		}
+
+		// If the file is not valid skip it
+		if !packageDecl.Valid() {
+			return nil
```

Comment:
Not sure if we have the infrastructure (and if not, maybe not in this PR) but this feels like a point where we could emit a warning, or something, so that people know if we are skipping over files they think we should be paying attention to. My sense is that you only want errors for files that are lock files but invalid, so perhaps we check to see if the error isn't "no version specified" and only raise then -- anything else is probably just random YAML. Maybe.

Reply:
I think we can pass in a diagnostic stream here, I'll have a look.